### ISSUE TITLE
推理时forward提取audio是在哪里？

### ISSUE THREAD

**Issue by fanbooo** [DAMO-NLP-SG/Video-LLaMA#56]
感谢大佬的工作，有两个问题想请教下；
1. 请问下forward里输入的只有video的帧数，img_embeds, atts_img = self.encode_audioQformer(image, modality_type=ModalityType.VISION) 的输入应该是图片batch？audio特征是怎么提取的呢？
2.   这段代码是图像特征还是图像和audio的融合特征呢?是与input_ids（groundtruth？）列拼接吗?
for cur_input_ids, cur_input_embeds in zip(input_ids, temp_input_embedding) 
      ....

def forward(self, samples):
        if 'conv_type' in samples.keys() and samples['conv_type']=='multi':
            
            im_patch_token_id = self.IMAGE_PATCH_TOKEN_ID
            image = samples["images"]
            input_ids = samples['input_ids']
            if len(image.size())==4:
                time = 1
                image = einops.repeat(image, 'b c h w -> b c t h w',t = time)

            if self.train_flag == 0:
                num_patch_tokens = self.num_video_query_token
                img_embeds, atts_img = self.encode_videoQformer_visual(image)
            elif self.train_flag == 1:
                num_patch_tokens = self.num_audio_query_token
                image = einops.rearrange(image, 'b c t h w -> b t c h w')
                img_embeds, atts_img = self.encode_audioQformer(image, modality_type=ModalityType.VISION)
                
            temp_input_ids = copy.deepcopy(input_ids)
            temp_input_ids[temp_input_ids == im_patch_token_id] = 0
            temp_input_embedding = self.llama_model.model.embed_tokens(temp_input_ids)

            new_input_embeds=[]
            cur_image_idx = 0
            for cur_input_ids, cur_input_embeds in zip(input_ids, temp_input_embedding):
                cur_image_features = img_embeds[cur_image_idx]

                if (cur_input_ids == im_patch_token_id).sum() != num_patch_tokens:
                        raise ValueError("The number of image patch tokens should be the same as the number of image patches.")
                masked_indices = torch.where(cur_input_ids == im_patch_token_id)[0]
                mask_index_start = masked_indices[0]
                if (masked_indices != torch.arange(mask_index_start, mask_index_start+num_patch_tokens, device=masked_indices.device, dtype=masked_indices.dtype)).any():
                    raise ValueError("The image patch tokens should be consecutive.")
                
                cur_new_input_embeds = torch.cat((cur_input_embeds[:mask_index_start], cur_image_features, cur_input_embeds[mask_index_start+num_patch_tokens:]), dim=0)
                new_input_embeds.append(cur_new_input_embeds)
                
                cur_image_idx+=1
            inputs_embeds = torch.stack(new_input_embeds, dim=0)
            targets = samples['labels']
            attention_mask = samples['attention_mask']
            with self.maybe_autocast():
                outputs = self.llama_model(
                    inputs_embeds=inputs_embeds,
                    attention_mask=attention_mask,
                    return_dict=True,
                    labels=targets,
                )
            loss = outputs.loss
            return {"loss": loss}

**Comment by hangzhang-nlp** [comment]
在训练的时候，我们并没有用audio-text的数据，而是用vision-text的数据来训练audio的适配器，将frozen的imagebind的输出空间（audio，vision共享）映射到LLM的输入空间，从而间接地达到audio-text的对齐效果。

forward函数只在训练的时候会调用，因此你看的这段代码只接受vision输入，融合的时候是将vision的特征和text的embedding 连接起来。即这句话：
```
cur_new_input_embeds = torch.cat((cur_input_embeds[:mask_index_start], cur_image_features, cur_input_embeds[mask_index_start+num_patch_tokens:]), dim=0)
```
提取audio和编码可以看这里，https://github.com/DAMO-NLP-SG/Video-LLaMA/blob/main/video_llama/conversation/conversation_video.py#L211 和https://github.com/DAMO-NLP-SG/Video-LLaMA/blob/main/video_llama/conversation/conversation_video.py#L220

**Comment by fanbooo** [comment]
谢谢大佬
